npm_and_yarn/lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb

```diff
@@ -214,10 +214,6 @@ def registry_source_url
           sources = dependency.requirements.
                     map { |r| r.fetch(:source) }.uniq.compact
 
-          # If there are multiple source types, or multiple source URLs, then
-          # it's unclear how we should proceed
-          raise "Multiple sources! #{sources.join(', ')}" if sources.map { |s| [s[:type], s[:url]] }.uniq.count > 1
```

Comment:
Does this actually raise atm with the test case? Looks like the type + url would be unique?

Reply:
I thought I wrote the failing test first but I'll double check. Thanks for asking.

Reply:
If I bring back the line above then the following test fails:

```bash
[dependabot-core-dev] ~/dependabot-core/npm_and_yarn $ bundle exec rspec spec/dependabot/npm_and_yarn/update_checker_spec.rb:228
Run options: include {:locations=>{"./spec/dependabot/npm_and_yarn/update_checker_spec.rb"=>[228]}}

Randomized with seed 3302
WARNING: Using `expect { }.not_to raise_error(SpecificErrorClass)` risks false positives, since literally any other error would cause the expectation to pass, including those raised by Ruby (e.g. `NoMethodError`, `NameError` and `ArgumentError`), meaning the code you are intending to test may not even get reached. Instead consider using `expect { }.not_to raise_error` or `expect { }.to raise_error(DifferentSpecificErrorClass)`. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /home/dependabot/dependabot-core/npm_and_yarn/spec/dependabot/npm_and_yarn/update_checker_spec.rb:228:in `block (4 levels) in <top (required)>'.
F

Failures:

  1) Dependabot::NpmAndYarn::UpdateChecker#latest_version with multiple requirements is expected not to raise /Multiple sources/
     Failure/Error: specify { expect { subject }.not_to raise_error(/Multiple sources/) }

       expected no /Multiple sources/, got #<RuntimeError: Multiple sources! {:type=>"registry", :url=>"https://registry.npmjs.org"}, {:type=>"registry", :url=>"https://registry.yarnpkg.com"}> with backtrace:
         # ./lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:217:in `registry_source_url'
         # ./lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:65:in `registry_url'
         # ./lib/dependabot/npm_and_yarn/update_checker/registry_finder.rb:35:in `dependency_url'
         # ./lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb:373:in `dependency_url'
         # ./lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb:302:in `fetch_npm_response'
         # ./lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb:283:in `npm_details'
         # ./lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb:104:in `valid_npm_details?'
         # ./lib/dependabot/npm_and_yarn/update_checker/latest_version_finder.rb:35:in `latest_version_from_registry'
         # ./lib/dependabot/npm_and_yarn/update_checker.rb:180:in `latest_released_version'
         # ./lib/dependabot/npm_and_yarn/update_checker.rb:205:in `latest_version_details'
         # ./lib/dependabot/npm_and_yarn/update_checker.rb:23:in `latest_version'
         # ./spec/dependabot/npm_and_yarn/update_checker_spec.rb:185:in `block (3 levels) in <top (required)>'
         # ./spec/dependabot/npm_and_yarn/update_checker_spec.rb:228:in `block (5 levels) in <top (required)>'
         # ./spec/dependabot/npm_and_yarn/update_checker_spec.rb:228:in `block (4 levels) in <top (required)>'
         # /home/dependabot/.bundle/ruby/2.6.0/gems/webmock-3.12.2/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'
     # ./spec/dependabot/npm_and_yarn/update_checker_spec.rb:228:in `block (4 levels) in <top (required)>'
     # /home/dependabot/.bundle/ruby/2.6.0/gems/webmock-3.12.2/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 0.02905 seconds (files took 1.36 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/dependabot/npm_and_yarn/update_checker_spec.rb:228 # Dependabot::NpmAndYarn::UpdateChecker#latest_version with multiple requirements is expected not to raise /Multiple sources/
```

It looks like I added a test to `update_checker_spec.rb` but not to `update_checker/registry_finder_spec.rb`. I will add one to the file that directly tests this class. Thanks @feelepxyz.

Reply:
Test case added [here](https://github.com/dependabot/dependabot-core/pull/3718/commits/25d5e756441c9a59e457086564d958bdcf50bcab).